interface/default.py

```diff
@@ -0,0 +1,96 @@
+import dis
+import warnings
+
+from .compat import PY3
+from .formatting import bulleted_list
+from .functional import keysorted, sliding_window
+
+
+class default(object):
+    """Default implementation of a function in terms of interface methods.
+    """
+    def __init__(self, implementation):
+        self.implementation = implementation
+
+    def __repr__(self):
+        return "{}({})".format(type(self).__name__, self.implementation)
+
+
+class DefaultUsesNonInterfaceMembers(UserWarning):
+    pass
+
+
+if PY3:  # pragma: nocover-py2
+    _DEFAULT_USES_NON_INTERFACE_MEMBER_TEMPLATE = (
+        "Default implementation of {iface}.{method} uses non-interface attributes.\n\n"
+        "The following attributes may be accessed but are not part of "
+        "the interface:\n"
+        "{non_members}\n\n"
+        "Consider changing the implementation of {method} or making these attributes"
+        " part of {iface}."
+    )
+
+    def warn_if_defaults_use_non_interface_members(interface_name,
+                                                   defaults,
+                                                   members):
+        """Warn if an interface default uses non-interface members of self.
+        """
+        for method_name, attrs in non_member_attributes(defaults, members):
+            warnings.warn(_DEFAULT_USES_NON_INTERFACE_MEMBER_TEMPLATE.format(
+                iface=interface_name,
+                method=method_name,
+                non_members=bulleted_list(attrs),
+            ))
+
+    def non_member_attributes(defaults, members):
+        from .typed_signature import TypedSignature
+
+        for default_name, default in keysorted(defaults):
+            impl = default.implementation
+
+            if isinstance(impl, staticmethod):
+                # staticmethods can't use attributes of the interface.
+                continue
+
+            self_name = TypedSignature(impl).first_argument_name
```

Comment:
if this is a property, we need to check `__get__`, `__set__`, and `__delete__`, right?

Reply:
In general, we only look at `fget` on properties right now.  I could see an argument for expanding to look at other fields, but I think that's a separate PR